rosetta/retriever/retriever.go

```diff
@@ -174,89 +173,14 @@ func (r *Retriever) Block(id identifier.Block) (*object.Block, []identifier.Tran
 
 	// Next, we step through all the transactions and accumulate events by transaction ID.
 	// NOTE: We consider transactions that don't generate any fund movements as irrelevant for now.
-	buckets := make(map[flow.Identifier][]object.Operation)
-	for _, event := range events {
-
-		// Decode the event payload into a Cadence value and cast to Cadence event.
-		value, err := json.Decode(event.Payload)
-		if err != nil {
-			return nil, nil, fmt.Errorf("could not decode event: %w", err)
-		}
-		e, ok := value.(cadence.Event)
-		if !ok {
-			return nil, nil, fmt.Errorf("could not cast event: %w", err)
-		}
-
-		// Check we have the necessary amount of fields.
-		if len(e.Fields) != 2 {
-			return nil, nil, fmt.Errorf("invalid number of fields (want: %d, have: %d)", 2, len(e.Fields))
-		}
-
-		// Now we have access to the fields for the events; the first one is always
-		// the amount, the second one the address. The types coming from Cadence
-		// are not native Flow types, so we need to use primitive types first.
-		vAmount := e.Fields[0].ToGoValue()
-		uAmount, ok := vAmount.(uint64)
-		if !ok {
-			return nil, nil, fmt.Errorf("could not cast amount (%T)", vAmount)
-		}
-		vAddress := e.Fields[1].ToGoValue()
-		bAddress, ok := vAddress.([flow.AddressLength]byte)
-		if !ok {
-			return nil, nil, fmt.Errorf("could not cast address (%T)", vAddress)
-		}
-
-		// Then, we can convert the amount to a signed integer so we can invert
-		// it and the address to a native Flow address.
-		amount := int64(uAmount)
-		address := flow.Address(bAddress)
-
-		// For the witdrawal event, we invert the amount into a negative number.
-		if event.Type == flow.EventType(withdrawal) {
-			amount = -amount
-		}
-
-		// Now we have everything to assemble the respective operation.
-		op := object.Operation{
-			ID: identifier.Operation{
-				Index: uint(event.EventIndex),
-			},
-			RelatedIDs: nil,
-			Type:       "TRANSFER",
-			Status:     "COMPLETED",
-			AccountID: identifier.Account{
-				Address: address.String(),
-			},
-			Amount: object.Amount{
-				Value: strconv.FormatInt(amount, 10),
-				Currency: identifier.Currency{
-					Symbol:   dps.FlowSymbol,
-					Decimals: dps.FlowDecimals,
-				},
-			},
-		}
-
-		// We store all operations for a transaction together in a bucket.
-		buckets[event.TransactionID] = append(buckets[event.TransactionID], op)
+	tmap, err := convert.EventsToTransactions(events, withdrawal)
+	if err != nil {
+		return nil, nil, err
```

Comment:
I would prefer seeing this:

```go
for _, event := range events {
  op, ok, err := convert.EventToOperation(event)
  if err != nil {
    return nil, nil, fmt.Errorf("could not convert event: %w", err)
  }
  if !ok {
    continue
  }
  // create tx if necessary and append op
}
```

The part about creating transactions based on which events are present will be changed anyway. We will first retrieve all Flow transactions for a block and create a Rosetta transaction for each. Then, we can append all successfully converted operations accordingly. This will fix the issue that we currently don't return transactions with no operations, and the issue that we return a transaction on the transaction route even if the given block doesn't contain it.